internal/driver/checkstatuses.go

```diff
@@ -0,0 +1,167 @@
+	devClient, edgexErr := d.newTemporaryOnvifClient(device)
+	if edgexErr != nil {
+		d.lc.Debugf("Connection to %s failed when creating client: %s", device.Name, edgexErr.Message())
+		// onvif connection failed, so lets probe it
+		if d.tcpProbe(device) {
+			return Reachable
+		}
+		return Unreachable
+
+	}
+
+	// sends get device information command to device (requires credentials)
+	_, edgexErr = devClient.callOnvifFunction(onvif.DeviceWebService, onvif.GetDeviceInformation, []byte{})
+	if edgexErr != nil {
+		d.lc.Debugf("%s command failed for device %s when using authentication: %s", onvif.GetDeviceInformation, device.Name, edgexErr.Message())
+		return UpWithoutAuth
+	}
+
+	return UpWithAuth
+}
+
+// // After failing to get a connection using authentication, it calls this function
+// // to try to reach the camera using a command that doesn't require authorization,
+// // and return a bool indicating success or failure
+// func (d *Driver) testConnectionNoAuth(device sdkModel.Device) bool {
+// 	// sends get capabilities command to device (does not require credentials)
+// 	// client, edgexErr := d.newTemporaryOnvifClient(device)
+// 	// client.callOnvifFunction(onvifdevice, onvifdevice.GetSystemDateAndTime, nil)
+// 	if edgexErr != nil {
+// 		d.lc.Debugf("Connection to %s failed when not using authentication: %s", device.Name, edgexErr.Message())
+// 		return false
+// 	}
+// 	return true
+// }
+
+// tcpProbe attempts to make a connection to a specific ip and port list to determine
+// if there is a service listening at that ip+port.
+func (d *Driver) tcpProbe(device sdkModel.Device) bool {
+	proto, ok := device.Protocols[OnvifProtocol]
+	if !ok {
+		d.lc.Warnf("Device %s is missing required %s protocol info, cannot send probe.", device.Name, OnvifProtocol)
+		return false
+	}
+	addr := proto[Address]
+	port := proto[Port]
+
+	if addr == "" || port == "" {
+		d.lc.Warnf("Device %s has no network address, cannot send probe.", device.Name)
+		return false
+	}
+	host := addr + ":" + port
+
+	conn, err := net.DialTimeout("tcp", host, time.Duration(d.config.AppCustom.ProbeTimeoutMillis)*time.Millisecond)
+	if err != nil {
+		d.lc.Debugf("Connection to %s failed when using simple tcp dial, Error: %s ", device.Name, err.Error())
+		return false
+	}
+	defer conn.Close()
+	return true
+}
+
+func (d *Driver) updateDeviceStatus(device sdkModel.Device, status string) error {
+	// todo: maybe have connection levels known as ints, so that way we can log at different levels based on
+	oldStatus := device.Protocols[OnvifProtocol][DeviceStatus]
+	if oldStatus != status {
+		d.lc.Infof("Device status for %s is now %s (used to be %s)", device.Name, status, oldStatus)
+	}
+
+	device.Protocols[OnvifProtocol][DeviceStatus] = status
+
+	if status != Unreachable {
+		device.Protocols[OnvifProtocol][LastSeen] = time.Now().Format(time.UnixDate)
+	}
+
+	return service.RunningService().UpdateDevice(device)
```

Comment:
maybe only update if something changed
```suggestion
	shouldUpdate := false
	
	oldStatus := device.Protocols[OnvifProtocol][DeviceStatus]
	if oldStatus != status {
		d.lc.Infof("Device status for %s is now %s (used to be %s)", device.Name, status, oldStatus)
		device.Protocols[OnvifProtocol][DeviceStatus] = status
		shouldUpdate = true
	}

	if status != Unreachable {
		device.Protocols[OnvifProtocol][LastSeen] = time.Now().Format(time.UnixDate)
		shouldUpdate = true
	}
	
	if shouldUpdate {
		return service.RunningService().UpdateDevice(device)
	}
	return nil
```